codeface/R/cluster/graph_comparison.r

```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
   g.2.intersect <- graph.adjacency(adj.matrix.2.intersect, mode = "directed")
 
-  graph.diff <- graph.difference(g.1.intersect, g.2.intersect)
+  graph.diff <- graph.difference(g.1.intersect, g.2.intersect, weighted = weighted, symmetric = symmetric)
 
   return(graph.diff)
 }
 
+get.index.map1 <- function(ids) {
+  N        <- nrow(ids)
+  map      <- new.env(size=N)
+  for(i in 1:N) {
+    map[[as.character(ids$id1[i])]] <- ids$id[i]
+  }
+  return(map)
+}
 
+get.index.map2 <- function(ids) {
+  N        <- nrow(ids)
+  map      <- new.env(size=N)
+  for(i in 1:N) {
+    map[[as.character(ids$id2[i])]] <- ids$id[i]
+  }
+  return(map)
+}
 ################################################################################
 ## High Level Functions
 ################################################################################
-run.graph.comparison <- function(con, pid.1, range.id.1, pid.2, range.id.2) {
+run.graph.comparison <- function(con, range.id.1, range.id.2, weighted=FALSE, symmetric=FALSE) {
   ## TODO: cluster method should not actually be required, we only need the main
   ##       graph, unfortunatley the database scheme currently marries the graph
   ##       to a clustering method
   cluster.method="Spin Glass Community"
+  pid.1 <- get.project.id.from.release.range.id(con, range.id.1)
+  pid.2 <- get.project.id.from.release.range.id(con, range.id.2)
   graph.data.1 <- get.graph.data.local(con, pid.1, range.id.1, cluster.method)
   graph.data.2 <- get.graph.data.local(con, pid.2, range.id.2, cluster.method)
-  edgelist.1   <- graph.data.1$edgelist
-  edgelist.2   <- graph.data.2$edgelist
+  # get the raw edgelists with the ids from the database
+  edgelist.1   <- graph.data.1$edgelist.db
+  edgelist.2   <- graph.data.2$edgelist.db
+
+  # Get the database ids and resolve them to names
   global.ids.1 <- graph.data.1$v.global.ids
   node.label.1 <- sapply(global.ids.1, function(id)
         query.person.name(con, id))
-  vertex.df.1 <- data.frame(name=global.ids.1, Id=node.label.1)
   global.ids.2 <- graph.data.2$v.global.ids
   node.label.2 <- sapply(global.ids.2, function(id)
         query.person.name(con, id))
-  vertex.df.2 <- data.frame(name=global.ids.2, Id=node.label.2)
 
-  ## Create igraph objects
-  g.1 <- graph.data.frame(edgelist.1, vertices=vertex.df.1, directed=TRUE)
-  g.2 <- graph.data.frame(edgelist.2, vertices=vertex.df.2, directed=TRUE)
+  # Create two dataframes to be merged by developer name
+  temp.vertex.df.1 <- data.frame(id1=global.ids.1, name=node.label.1)
+  temp.vertex.df.2 <- data.frame(id2=global.ids.2, name=node.label.2)
+
+  # Merge the graphs by developers and create a mapping to local ids
+  temp.merged <- merge (temp.vertex.df.1, temp.vertex.df.2, all=T, by="name")
+  merged.vertex <- data.frame(id=1:nrow(temp.merged), name=temp.merged$name, id1=temp.merged$id1, id2=temp.merged$id2)
+  map1 <- get.index.map1(merged.vertex)
+  map2 <- get.index.map2(merged.vertex)
```

Comment:
Here I try to merge the two dataframes `temp.vertex.df.1` and `2` which contain the ids of the developers (same developer can have different ids) into a new dataframe `temp.merged`. And then I create local ids for the merged frame. From `merged.vertex` we can then create two mappings (one for `edgelist.1` and one for `edgelist.2`) to map the global developer ids into our new local id.